src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultCollectionValidationStrategy.cs

```diff
@@ -60,9 +63,16 @@ private DefaultCollectionValidationStrategy()
             return new Enumerator(metadata.ElementMetadata, key, enumerator);
         }
 
-        public static IEnumerator GetEnumeratorForElementType(ModelMetadata metadata, object model)
+        public IEnumerator GetEnumeratorForElementType(ModelMetadata metadata, object model)
         {
-            var getEnumeratorMethod = _getEnumerator.MakeGenericMethod(metadata.ElementType);
+            MethodInfo getEnumeratorMethod;
+            if (!_genericGetEnumeratorCache.TryGetValue(metadata.ElementType, out getEnumeratorMethod))
+            {
+                getEnumeratorMethod = _genericGetEnumeratorCache.AddOrUpdate(
+                        key: metadata.ElementType,
+                        addValueFactory: (type) => _getEnumerator.MakeGenericMethod(type),
+                        updateValueFactory: (type, methodInfo) => methodInfo);
+            }
             return (IEnumerator)getEnumeratorMethod.Invoke(null, new object[] { model });
```

Comment:
I'm currently testing the performance of changing this to an expression tree.

@rynowak I'm failing to understand if your comment is related to the cache we're introducing, or the overall design on how we invoke `GetEnumerator`. The code is only caching a call to a method that knows the type of `T` and operates on a live instance:
```C#
// Called via reflection.
private static IEnumerator GetEnumerator<T>(object model)
{
    return (model as IEnumerable<T>)?.GetEnumerator() ?? ((IEnumerable)model).GetEnumerator();
}
```

Are you saying that we can expect a situation in which the `metadata.ElementType` used to define `T` is inconsistent to the type obtained by `((IEnumerable)model).GetEnumerator()`? If so, is the cache affecting this situation?

Sorry if my questioning is too basic.